Added Causal Mask Pattern Fusion for LongRoPe Models

tadani3 opened the pull request:
Modification of the GQA causal mask fusion rule to handle the attention mask fusion for Longrope models such as Phi-4-mini-reasoning. The causal mask modification leads to a result that matches the optimizations made in ModelBuilder.

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
+        """
+        return (id(attention_mask))
+    
+    def compute_mask(self, op, attention_mask : _onnx_types.INT64['batch', 'seq_len']):
```
The rewriter doesn't use onnxscript type (yet). Could you instead use a comment to document the shape of the attention_mask?

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
             _outputs=3,
         )
 
+class LongRoPeGQACausalMask(pattern.RewriteRuleClassBase):
```
Could you use the docstring to document the pattern and its replacement? For the branches A, B, and C, I would consider giving them descriptive names.

gramalingam replied:
The following is my understanding: if this is correct, maybe they can be renamed appropriately: I believe that A constructs the kv_range, B constructs the query_range, and C constructs the batch_range. Each constructs the corresponding range as a 4D tensor with 1s in other position (for constructing a final attention-mask of shape [Batch, NumHeads, QueryRange, KVRange] via broadcast). 

I am a bit puzzled that query_range and kv_range look to be the same here, it might be an artifact of this model-usage, I guess.

tadani3 replied:
I wasn't sure what the branches referred to but I'll make changes following what Rama is suggesting.

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
+        total_seq_length,
+    ):
+        seq_len = op.Shape(input_ids, end=2, start=1, _outputs=["seq_len"])
+        seq_len_0D = op.Squeeze(seq_len, _outputs=["seq_len_0D"])
```
```suggestion
        seq_len_0d = op.Squeeze(seq_len, _outputs=["seq_len_0d"])
```

prefer snake case for variable names when possible

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
+        mask_A_B_combined = op.And(mask_A_B_greater_bitwise, mask_A_B_less, _outputs=["mask_A_B_combined"])
+        mask_A_B_combined_bitwise = op.And(True, mask_A_B_combined, _outputs=["mask_A_B_combined_bitwise"])
+
+        # EXPAND B/C TO AND
```
I would document the branches in plain English for readers

justinchuby reviewed (commented):
Thanks! Could you resolve the lint errors by running `lintrunner -a` and fix any issues?

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
+class LongRoPeGQACausalMask(pattern.RewriteRuleClassBase):
+    def __init__(self):
+        super().__init__("LongRoPeGQACausalMask", remove_nodes=False)
+        self._mask_cache = {}
```
The copilot message is reasonable: the rewrite rule class should be stateless.

gramalingam replied:
I think use of state for this purpose is okay? It has been used before for a similar purpose: which is to introduce values that are reused across multiple rewrites. (Now that we have CSE, there is an alternative path, which is to create duplicate copies and then eliminate them via CSE ... but I am not sure it is worth the bother.)

justinchuby replied:
My concern is that the states will transfer from model to another if not careful, which is probably not a good idea. Maybe we can have a class managed state dict that will be cleared by the class?

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
+class LongRoPeGQACausalMask(pattern.RewriteRuleClassBase):
+    def __init__(self):
+        super().__init__("LongRoPeGQACausalMask", remove_nodes=False)
+        self._mask_cache = {}
```
We should not maintain a state in the rule.

gramalingam commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
+        mask_expanded_C = op.Expand(reshaped_range_C, mask_shape_C_abs, _outputs=["mask_expanded_C"])
+
+        # EXPAND A/B TO AND
+        mask_expanded_A_sub = op.Sub(mask_expanded_A, 262144, _outputs=["mask_expanded_A_sub"])
```
Better to make it a pattern-variable, I think ... if I understand right, this is actually a magic sequence-length constant? Perhaps model-specific?

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
+        super().__init__("LongRoPeGQACausalMask", remove_nodes=False)
+        self._mask_cache = {}
+    
+    def _get_mask_key(self, attention_mask):
```
In general, avoid creating class methods that do not require states from self, and instead make them module-level private functions for testability and clarity.

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -354,9 +355,163 @@ def rewrite(
+    
+    def _get_mask_key(self, attention_mask):
+        """
+        Generate a unique key for the mask based on input_ids and past_kv_cache.
+        This is used to cache the mask to avoid recomputation.
+        """
+        return (id(attention_mask))
```
I agree _get_mask_key seems unecessary. We can use the Value objects directly as keys.

justinchuby commented on onnxscript/rewriter/ort_fusions/gqa.py:
```diff
@@ -7,6 +7,7 @@
 import numpy as np
 import onnx_ir as ir
 
+import onnxscript.onnx_types as _onnx_types
```
```suggestion
```

_onnx_types is incompatible with the rewriter (yet)

tadani3 commented:
@microsoft-github-policy-service agree company="Microsoft"

tadani3 commented:
I added a class called Phi4MiniReasoningPostProcessor which uses the ONNX IR to fulfill two tasks:

1. Inserts the cached Cos/Sin values used in the Rotary Embeddings

- The Cos/Sin caches are computed offline following the logic found in Transformer's modeling_phi3.py file.
- The existing pattern that computes Cos and Sin is found in the ONNX model graph and removed.
- An If node containing both sets of caches is inserted to the graph. The selection condition for the two sets of caches is the total_sequence_length, which we obtain from the attention_mask.
- GQA nodes within the graph are updated to have their cos/sin inputs come from the newly inserted caches.

2. Removes Position Ids and remaining child nodes from the model graph.

- Matches the position ids and following nodes that were used in the old Cos/Sin value computation.
- Uses the ONNX IR to remove these nodes from the graph.

Reasoning and Motivation

- These changes allow the version of Phi4-mini-reasoning produced using the DynamoExporter to match the version found in Model Builder

gramalingam commented:
I think this is outdated and not needed?
